### PR TITLE
Promises

### PR DESCRIPTION
**Pull request by dlongley**

### CODE REVIEW

File: lib/blocks.js [hunk 1]
```diff
@@ -7,242 +7,207 @@ const _ = require('lodash');
...
+        owner: key.owner,
+        publicKeyBase58: key.publicKeyBase58
+      });
+
+  // update events with block information 100 at a time
+  const hashes = consensusResult.eventHash;
+  let now = Date.now();
+  const limit = pLimit(100);
+  await Promise.all(hashes.map((eventHash, index) => {
+    limit(() => ledgerNode.storage.events.update({
```
Comment (davidlehn) [inline lib/blocks.js]: Looks like it should return the resulting promise from the limit(...) call, or just remove the surrounding `{...}`.

File: lib/election.js [hunk 2]
```diff
@@ -69,43 +57,46 @@ api._findMergeEventProof = api._consensus._findMergeEventProof;
...
+  let consensus;
+  let err;
+  try {
+    consensus = await _runConsensusInPool(
+      {ledgerNode, history, blockHeight, electors});
+  } finally {
+    // failure to cache logging info not critical, ignore in `.catch`
+    const duration = Date.now() - startTime;
+    cache.client.set(`findConsensus|${ledgerNode.id}`, duration)
+      .catch(e => () => {});
```
Comment (davidlehn) [inline lib/election.js]: `e => {}`

File: lib/events.js [hunk 3]
```diff
@@ -1559,106 +1148,229 @@ function _verifyGenesisEvents(
...
-    }
-    const {creator} = results.verify;
-    callback(null, {creator, mergeEvent, mergeHash});
-  });
+
+  // Note: `signature.verify` throws an error if verification fails
+  const {keyOwner: creator} = await signature.verify(
+    {doc: mergeEvent, ledgerNodeId});
+  const proofHash = await _util.hasher(proof);
+  const mergeHash = await _util.hasher(mergeEvent);
```
Comment (davidlehn) [inline lib/events.js]: Could run all three above calls in parallel as before.

File: lib/voters.js [hunk 4]
```diff
@@ -39,188 +40,136 @@ module.exports = api;
...
+      // lock acquired, create voter
+      voter = await _createVoter(ledgerNodeId);
+
+      // release lock
+      await cache.client.del(lockKey);
     }
-    callback(null, results.voter);
-  });
-};
+  }, {retries: 100});
```
Comment (davidlehn) [inline lib/voters.js]: The behavior of these p-retry params is *far* different than the previous async.retry!  p-retry is *exponential* and starts at a 1s delay vs the previous constant 50ms interval retries.  The last attempt here will be closer to never than 5s. ;-)
https://github.com/tim-kos/node-retry#retryoperationoptions
https://caolan.github.io/async/docs.html#retry

File: test/mocha/68-election-get-ancestors.js [hunk 5]
```diff
@@ -144,18 +141,19 @@ describe('Election API _getAncestors', () => {
...
-          assertNoError(err);
+        try {
+          const result = await getAncestors({hashes, ledgerNode});
           should.exist(result);
           result.should.be.an('array');
           result.should.have.length(2);
-          callback();
-        });
+        } catch(e) {
+          assertNoError(e);
```
Comment (davidlehn) [inline test/mocha/68-election-get-ancestors.js]: async func, can remove try/catch and let error propagate out.

File: test/mocha/68-election-get-ancestors.js [hunk 6]
```diff
@@ -169,21 +167,22 @@ describe('Election API _getAncestors', () => {
...
-          assertNoError(err);
+        try {
+          const result = await getAncestors({hashes, ledgerNode});
           should.exist(result);
           result.should.be.an('array');
           result.should.have.length(4);
-          callback();
-        });
+        } catch(e) {
+          assertNoError(e);
```
Comment (davidlehn) [inline test/mocha/68-election-get-ancestors.js]: remove try/catch

File: test/mocha/68-election-get-ancestors.js [hunk 7]
```diff
@@ -210,13 +209,14 @@ describe('Election API _getAncestors', () => {
...
```
Comment (davidlehn) [inline test/mocha/68-election-get-ancestors.js]: remove try/catch

File: test/mocha/68-election-get-ancestors.js [hunk 8]
```diff
@@ -240,20 +240,23 @@ describe('Election API _getAncestors', () => {
...
+          });
+          should.exist(result);
+          result.should.be.an('array');
+          result.should.have.length(4);
+          result.forEach(e => {
+            e.event.should.be.an('array');
+            e.event.should.have.length(1);
+          });
+        } catch(e) {
+          assertNoError(e);
```
Comment (davidlehn) [inline test/mocha/68-election-get-ancestors.js]: remove try/catch

File: test/mocha/68-election-get-ancestors.js (hunk 5 above)
Comment (dlongley) [reply test/mocha/68-election-get-ancestors.js]: I'm not touching the tests (as much as possible) in this pass.

**Comment by dlongley** [comment]
@davidlehn addressed your comments.

**Pull request merged**